app/invocation/invocation_error_card.tsx

```diff
@@ -3,75 +3,174 @@ import React from "react";
 import InvocationModel from "./invocation_model";
 import { failure_details } from "../../proto/failure_details_ts_proto";
 import TerminalComponent from "../terminal/terminal";
-import rpc_service from "../service/rpc_service";
+import rpc_service, { CancelablePromise } from "../service/rpc_service";
+import { exitCode } from "../util/exit_codes";
+import { build_event_stream } from "../../proto/build_event_stream_ts_proto";
 
 const debugMessage =
   "Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging";
+
 interface Props {
   model: InvocationModel;
 }
 
 interface State {
-  stdErr: string;
+  model?: CardModel;
 }
 
+type CardModel = {
+  errors: ErrorData[];
+};
+
+type ErrorData = {
+  action?: build_event_stream.ActionExecuted;
+  actionStderr?: string;
+  actionStdout?: string;
+
+  aborted?: build_event_stream.BuildEvent;
+
+  finished?: build_event_stream.BuildFinished;
+};
+
 /**
  * Displays a card containing the main reason that an invocation failed.
  */
 export default class ErrorCardComponent extends React.Component<Props, State> {
-  state: State = {
-    stdErr: "",
-  };
+  state: State = {};
 
   componentDidMount() {
-    if (this.props.model.failedAction?.action?.stderr?.uri) {
-      rpc_service
-        .fetchBytestreamFile(
-          this.props.model.failedAction?.action.stderr?.uri,
-          this.props.model.getInvocationId(),
-          "text"
-        )
-        .then((resp) => this.setState({ stdErr: resp }));
+    this.fetchModelData(this.getModel());
+  }
+
+  componentDidUpdate(prevProps: Props): void {
+    const model = this.getModel(this.props);
+    if (!modelsEqual(this.getModel(this.props), this.getModel(prevProps))) {
+      this.fetchModelData(model);
     }
   }
 
-  prettyPrintedStdErr() {
-    if (!this.state.stdErr) {
-      return "";
+  getModel(props = this.props): CardModel {
```

Comment:
small nit, i feel like defaulting is kinda error-prone here, and given that you only use it this way in `componentDidMount`, you might as well make this a standalone function